<a href="https://colab.research.google.com/github/crowley409/Deep-Learning-Challenge/blob/main/Deep_Learning_Challenge.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=["EIN", "NAME"])

In [3]:
# Determine the number of unique values in each column.
unique_value_counts = application_df.nunique()

print("Number of Unique Values in Each Column:")
print(unique_value_counts)

Number of Unique Values in Each Column:
APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64


In [4]:
# Look at APPLICATION_TYPE value counts for binning
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()

print("APPLICATION_TYPE Value Counts:")
print(application_type_counts)

APPLICATION_TYPE Value Counts:
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64


In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
cutoff_value = 500

application_types_to_replace = list(application_type_counts[application_type_counts < cutoff_value].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
print("Updated APPLICATION_TYPE Value Counts:")
print(application_df['APPLICATION_TYPE'].value_counts())

Updated APPLICATION_TYPE Value Counts:
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64


In [6]:
# Look at CLASSIFICATION value counts for binning
classification_counts = application_df['CLASSIFICATION'].value_counts()

print("CLASSIFICATION Value Counts:")
print(classification_counts)

CLASSIFICATION Value Counts:
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64


In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
classification_counts_gt1 = classification_counts[classification_counts > 1]

print("CLASSIFICATION Value Counts (>1):")
print(classification_counts_gt1)

CLASSIFICATION Value Counts (>1):
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64


In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classification_cutoff_value = 100

classifications_to_replace = list(classification_counts[classification_counts < classification_cutoff_value].index)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
print("Updated CLASSIFICATION Value Counts:")
print(application_df['CLASSIFICATION'].value_counts())

Updated CLASSIFICATION Value Counts:
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      669
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: CLASSIFICATION, dtype: int64


In [9]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df, columns=["APPLICATION_TYPE", "AFFILIATION", "CLASSIFICATION", "USE_CASE", "ORGANIZATION", "INCOME_AMT", "SPECIAL_CONSIDERATIONS"])

print("Updated DataFrame after One-Hot Encoding:")
print(application_df.head())

Updated DataFrame after One-Hot Encoding:
   STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                       0   
1       1   108590              1                       0   
2       1     5000              0                       0   
3       1     6692              1                       0   
4       1   142590              1                       0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                     1                     0                    0   
1                     0                     0                    1   
2                     0                     0                    0   
3                     0                     0                    1   
4                     0                     0                    1   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                    0                    0                    0  ...   
1                    0  

In [10]:
# Split our preprocessed data into our features and target arrays
X = application_df.drop(columns="IS_SUCCESSFUL")
y = application_df["IS_SUCCESSFUL"]

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify=y)

In [11]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [13]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=8, activation='relu', input_dim=len(X_train.columns)))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=5, activation='relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 8)                 400       
                                                                 
 dense_4 (Dense)             (None, 5)                 45        
                                                                 
 dense_5 (Dense)             (None, 1)                 6         
                                                                 
Total params: 451 (1.76 KB)
Trainable params: 451 (1.76 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [14]:
# Compile the model
nn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [15]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=10, shuffle=True, verbose=2)

Epoch 1/10
804/804 - 3s - loss: 0.6054 - accuracy: 0.6930 - 3s/epoch - 4ms/step
Epoch 2/10
804/804 - 1s - loss: 0.5654 - accuracy: 0.7261 - 1s/epoch - 2ms/step
Epoch 3/10
804/804 - 1s - loss: 0.5598 - accuracy: 0.7277 - 1s/epoch - 2ms/step
Epoch 4/10
804/804 - 1s - loss: 0.5570 - accuracy: 0.7290 - 1s/epoch - 2ms/step
Epoch 5/10
804/804 - 1s - loss: 0.5554 - accuracy: 0.7296 - 1s/epoch - 2ms/step
Epoch 6/10
804/804 - 1s - loss: 0.5538 - accuracy: 0.7310 - 1s/epoch - 2ms/step
Epoch 7/10
804/804 - 1s - loss: 0.5519 - accuracy: 0.7329 - 1s/epoch - 2ms/step
Epoch 8/10
804/804 - 1s - loss: 0.5515 - accuracy: 0.7323 - 1s/epoch - 2ms/step
Epoch 9/10
804/804 - 1s - loss: 0.5507 - accuracy: 0.7322 - 1s/epoch - 2ms/step
Epoch 10/10
804/804 - 1s - loss: 0.5496 - accuracy: 0.7328 - 1s/epoch - 2ms/step


In [16]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5580 - accuracy: 0.7234 - 999ms/epoch - 4ms/step
Loss: 0.5580171346664429, Accuracy: 0.7233819365501404


In [17]:
# Export our model to HDF5 file
nn.save("AlphabetSoutCharity.h5")

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3079: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [18]:
# Start of optimization
optimized_nn = tf.keras.models.Sequential()
optimized_nn.add(tf.keras.layers.Dense(units=16, activation='relu', input_dim=len(X_train.columns)))
optimized_nn.add(tf.keras.layers.Dense(units=10, activation='relu'))
optimized_nn.add(tf.keras.layers.Dense(units=5, activation='relu'))
optimized_nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

In [19]:
# Compilation of optimized nn
optimized_nn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [20]:
# Optimized model training
optimized_fit_model = optimized_nn.fit(X_train_scaled, y_train, epochs=20, shuffle=True, verbose=2, validation_data=(X_test_scaled, y_test))

Epoch 1/20
804/804 - 3s - loss: 0.6120 - accuracy: 0.7041 - val_loss: 0.5765 - val_accuracy: 0.7198 - 3s/epoch - 4ms/step
Epoch 2/20
804/804 - 2s - loss: 0.5631 - accuracy: 0.7289 - val_loss: 0.5625 - val_accuracy: 0.7209 - 2s/epoch - 2ms/step
Epoch 3/20
804/804 - 2s - loss: 0.5537 - accuracy: 0.7301 - val_loss: 0.5635 - val_accuracy: 0.7187 - 2s/epoch - 2ms/step
Epoch 4/20
804/804 - 2s - loss: 0.5515 - accuracy: 0.7316 - val_loss: 0.5588 - val_accuracy: 0.7195 - 2s/epoch - 3ms/step
Epoch 5/20
804/804 - 2s - loss: 0.5498 - accuracy: 0.7325 - val_loss: 0.5575 - val_accuracy: 0.7221 - 2s/epoch - 3ms/step
Epoch 6/20
804/804 - 2s - loss: 0.5488 - accuracy: 0.7341 - val_loss: 0.5580 - val_accuracy: 0.7235 - 2s/epoch - 2ms/step
Epoch 7/20
804/804 - 2s - loss: 0.5474 - accuracy: 0.7346 - val_loss: 0.5575 - val_accuracy: 0.7219 - 2s/epoch - 2ms/step
Epoch 8/20
804/804 - 2s - loss: 0.5471 - accuracy: 0.7347 - val_loss: 0.5575 - val_accuracy: 0.7236 - 2s/epoch - 2ms/step
Epoch 9/20
804/804 - 2s 

In [21]:
# Evaluate the model using the optimized data
model_loss, model_accuracy = optimized_nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5584 - accuracy: 0.7245 - 382ms/epoch - 1ms/step
Loss: 0.5584132075309753, Accuracy: 0.7245481014251709


In [22]:
from tensorflow.keras.layers import BatchNormalization, Dropout

# Optimized model with batch normalization and dropout
optimized_nn_2 = tf.keras.models.Sequential()
optimized_nn_2.add(tf.keras.layers.Dense(units=16, activation='relu', input_dim=len(X_train.columns)))
optimized_nn_2.add(BatchNormalization())
optimized_nn_2.add(Dropout(0.5))
optimized_nn_2.add(tf.keras.layers.Dense(units=10, activation='relu'))
optimized_nn_2.add(BatchNormalization())
optimized_nn_2.add(Dropout(0.5))
optimized_nn_2.add(tf.keras.layers.Dense(units=5, activation='relu'))
optimized_nn_2.add(BatchNormalization())
optimized_nn_2.add(Dropout(0.5))
optimized_nn_2.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

In [23]:
# Compile
optimized_nn_2.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [24]:
# Train
optimized_fit_model_2 = optimized_nn_2.fit(X_train_scaled, y_train, epochs=20, shuffle=True, verbose=2, validation_data=(X_test_scaled, y_test))

Epoch 1/20
804/804 - 5s - loss: 0.7354 - accuracy: 0.5648 - val_loss: 0.6453 - val_accuracy: 0.6675 - 5s/epoch - 6ms/step
Epoch 2/20
804/804 - 3s - loss: 0.6607 - accuracy: 0.6167 - val_loss: 0.6211 - val_accuracy: 0.7003 - 3s/epoch - 4ms/step
Epoch 3/20
804/804 - 3s - loss: 0.6362 - accuracy: 0.6593 - val_loss: 0.6055 - val_accuracy: 0.7040 - 3s/epoch - 4ms/step
Epoch 4/20
804/804 - 2s - loss: 0.6270 - accuracy: 0.6714 - val_loss: 0.5951 - val_accuracy: 0.7099 - 2s/epoch - 3ms/step
Epoch 5/20
804/804 - 2s - loss: 0.6203 - accuracy: 0.6848 - val_loss: 0.5885 - val_accuracy: 0.7167 - 2s/epoch - 3ms/step
Epoch 6/20
804/804 - 3s - loss: 0.6147 - accuracy: 0.6930 - val_loss: 0.5853 - val_accuracy: 0.7159 - 3s/epoch - 3ms/step
Epoch 7/20
804/804 - 3s - loss: 0.6089 - accuracy: 0.7023 - val_loss: 0.5796 - val_accuracy: 0.7205 - 3s/epoch - 4ms/step
Epoch 8/20
804/804 - 3s - loss: 0.6048 - accuracy: 0.7056 - val_loss: 0.5778 - val_accuracy: 0.7216 - 3s/epoch - 4ms/step
Epoch 9/20
804/804 - 3s 

In [25]:
# Evaluate the model using the newly optimized data
model_loss, model_accuracy = optimized_nn_2.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5669 - accuracy: 0.7224 - 406ms/epoch - 2ms/step
Loss: 0.5668942928314209, Accuracy: 0.722449004650116


In [26]:
# Simpler optimized model
optimized_nn_3 = tf.keras.models.Sequential()
optimized_nn_3.add(tf.keras.layers.Dense(units=8, activation='relu', input_dim=len(X_train.columns)))
optimized_nn_3.add(tf.keras.layers.Dense(units=4, activation='relu'))
optimized_nn_3.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

In [27]:
# Compile with lower learning rate
optimizer = tf.keras.optimizers.Adam(learning_rate=0.0005)
optimized_nn_3.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])

In [28]:
# Train with increased epochs
optimized_fit_model_3 = optimized_nn_3.fit(X_train_scaled, y_train, epochs=40, shuffle=True, verbose=2, validation_data=(X_test_scaled, y_test))

Epoch 1/40
804/804 - 3s - loss: 0.6683 - accuracy: 0.6475 - val_loss: 0.6256 - val_accuracy: 0.6988 - 3s/epoch - 4ms/step
Epoch 2/40
804/804 - 2s - loss: 0.5880 - accuracy: 0.7223 - val_loss: 0.5887 - val_accuracy: 0.7127 - 2s/epoch - 2ms/step
Epoch 3/40
804/804 - 2s - loss: 0.5681 - accuracy: 0.7247 - val_loss: 0.5785 - val_accuracy: 0.7170 - 2s/epoch - 2ms/step
Epoch 4/40
804/804 - 2s - loss: 0.5612 - accuracy: 0.7274 - val_loss: 0.5744 - val_accuracy: 0.7205 - 2s/epoch - 2ms/step
Epoch 5/40
804/804 - 2s - loss: 0.5578 - accuracy: 0.7285 - val_loss: 0.5719 - val_accuracy: 0.7214 - 2s/epoch - 2ms/step
Epoch 6/40
804/804 - 3s - loss: 0.5561 - accuracy: 0.7296 - val_loss: 0.5709 - val_accuracy: 0.7217 - 3s/epoch - 3ms/step
Epoch 7/40
804/804 - 2s - loss: 0.5544 - accuracy: 0.7296 - val_loss: 0.5696 - val_accuracy: 0.7216 - 2s/epoch - 3ms/step
Epoch 8/40
804/804 - 2s - loss: 0.5534 - accuracy: 0.7301 - val_loss: 0.5673 - val_accuracy: 0.7222 - 2s/epoch - 2ms/step
Epoch 9/40
804/804 - 2s 

In [31]:
# Evaluate the model using the simplified optimized data
model_loss, model_accuracy = optimized_nn_3.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5571 - accuracy: 0.7251 - 1s/epoch - 4ms/step
Loss: 0.5570860505104065, Accuracy: 0.7251312136650085


1. Overview:

*   The primary goal of this analysis was to build a neural network model using TensorFlow to predict whether charities who applied for funding from Alphabet Soup would be successful. The dataset provided valuable information about each applicant to help in making this assessment.

---
2. Results:

*   The target variable for this assessment was the IS_SUCCESSFUL column
*   The feature variables were all of the other columns in the dataset, except for EIN and NAME which were excluded
*   The initial neural network included three layers: An input layer, a hidden layer with 8 neurons and relu activation, and an output layer with 1 neuron and sigmoid activation. The model was compiled using the Adam optimizer and binary cross-entropy loss function.
*   The initial model reached an accuracy of about 72%, which did not meet the requirement of 75%.
*   Optimization attempts were made by making edits such as adding more neurons, hidden layers, different activation functions, and adjusting parameters.
*   Despite these efforts, accuracy did not significantly improve.

---
3. Summary:

*   While the deep learning model was able to reach a reasonable accuracy, it still fell short of the target goal of 75%. Since further optimization attempts did not yield significant improvements, it would be recommended to explore other forms of machine learning algorithms such as Random Forest, which may be better equipped to capture complex patterns in the data more efficiently.




